app/design/adminhtml/default/default/template/promo/salesrulejs.phtml

```diff
@@ -5,104 +5,94 @@
-            'select', 'textarea').collect( function(elm) {
-        return Validation.validate(elm, {
-            useTitle :false,
-            onElementValidate : function() {
-            }
+async function generateCouponCodes(idPrefix, generateUrl, grid) {
+    const fieldsetEl = document.getElementById(idPrefix + 'information_fieldset');
+    const elements = fieldsetEl.querySelectorAll('input, select, textarea');
+    let validationResult = true;
+
+    fieldsetEl.classList.remove('ignore-validate');
+    for (const el of elements) {
+        validationResult &&= Validation.validate(el, {
+            useTitle: false,
+            onElementValidate: function() {},
         });
-    }).all();
-    $(idPrefix + 'information_fieldset').addClassName('ignore-validate');
+    }
+    fieldsetEl.classList.add('ignore-validate');
 
     if (!validationResult) {
         return;
     }
-    var elements = $(idPrefix + 'information_fieldset').select('input', 'select', 'textarea');
 
-    elements = elements.concat(
-        $$('#rule_uses_per_coupon'),
-        $$('#rule_uses_per_customer'),
-        $$('#rule_to_date')
-    );
+    const formData = new FormData();
+    for (const el of elements) {
+        formData.append(el.name, el.value);
+    }
+    for (const extra of ['rule_uses_per_coupon', 'rule_uses_per_customer', 'rule_to_date']) {
+        const el = document.getElementById(extra);
+        formData.append(el.name, el.value);
+    }
+
+    clearMessagesDiv();
 
-    var params = Form.serializeElements(elements, true);
-    params.form_key = FORM_KEY;
-    $('messages').update();
-    var couponCodesGrid = eval(grid);
-    new Ajax.Request(generateUrl, {
-        parameters :params,
-        method :'post',
-        onComplete : function (transport, param){
-            var response = transport.responseJSON || transport.responseText.evalJSON(true) || {};
+    try {
+        const result = await mahoFetch(generateUrl, {
+            method: 'POST',
+            body: formData,
+        });
 
-            if (couponCodesGrid) {
-                couponCodesGrid.reload();
-            }
-            if (response && response.messages) {
-                $('messages').update(response.messages);
-            }
-            if (response && response.error) {
-                alert(response.error.stripTags().toString());
-            }
+        if (window[grid]) {
+            window[grid].reload();
         }
-    });
-}
-
-Ajax.Responders.register({
-    onComplete: function() {
-        if ($('promo_catalog_edit_tabs_coupons_section_content')
-            && $('promo_catalog_edit_tabs_coupons_section_content').visible()
-            && Ajax.activeRequestCount == 0
-        ) {
-            handleCouponsTabContentActivity();
+        if (result?.messages) {
+            setMessagesDivHtml(result.messages);
         }
+    } catch (error) {
+        setMessagesDiv(error.message, 'error');
```

Comment:
```suggestion
        setMessagesDiv(error?.message || 'An error occurred', 'error');
```

does something like this make sense?

Reply:
Here it isn't needed, since error is a JS `Error` object which always has the message property.

But it could be useful to put it in `mahoFetch` in case a controller sends back only `{ error: true }`. I might as well add that to this PR.